packages/core/src/external/fhir/shared/index.ts

```diff
@@ -121,3 +127,50 @@ export function getIdFromSubjectRef(subject: Reference | undefined): string | un
   }
   return undefined;
 }
+
+function isPatient(resource: Resource | undefined): resource is Patient {
+  return resource?.resourceType === "Patient";
+}
+
+function isOrganization(resource: Resource | undefined): resource is Organization {
+  return resource?.resourceType === "Organization";
+}
+
+export function isComposition(resource: Resource | undefined): resource is Composition {
+  return resource?.resourceType === "Composition";
+}
+
+export function isObservation(resource: Resource | undefined): resource is Observation {
+  return resource?.resourceType === "Observation";
+}
+
+export function isDiagnosticReport(resource: Resource | undefined): resource is DiagnosticReport {
+  return resource?.resourceType === "DiagnosticReport";
+}
+
+export function findOrganizationResource(fhirBundle: Bundle): Organization | undefined {
+  const organizationEntry = fhirBundle.entry?.find(entry => isOrganization(entry.resource));
+  if (organizationEntry && isOrganization(organizationEntry.resource)) {
+    return organizationEntry.resource;
+  }
+  return undefined;
+}
+
+export function findPatientResource(fhirBundle: Bundle): Patient | undefined {
+  return fhirBundle.entry?.map(e => e.resource).find(isPatient);
+}
+
+export function findResourceInBundle(bundle: Bundle, reference: string): Resource | undefined {
+  if (!bundle.entry) {
+    return undefined;
+  }
+  const entry = bundle.entry.find(entry => {
+    const entryReference = buildEntryReference(entry.resource);
+    return entryReference === reference;
+  });
+  return entry?.resource;
+}
+
+export function buildEntryReference(resource: Resource | undefined): string {
+  return resource ? `${resource.resourceType}/${resource.id}` : "";
+}
```

Comment:
I think the function signature should be this: 
```export function buildEntryReference(resource: Resource): string {```

Reply:
That too. I understand why we would want to have `| undefined` as the input param, but either way, I'd have the response matching the input WRT to `undefined` or not` - we should avoid empty string to represent empty value like the plague. 😅

Reply:
OOC, why such a drastic response to that? :)

Reply:
Ahahaha... just trying to say I think we should avoid this as much as we can. :)

Imagine you code expecting a param as a string to build a URL. You get empty string, now we're calling a diff endpoint. Sure, we can say the function should validate the param is valid in the sense it meets not only the type constraints but it's also a valid value. But this adds on the complexity of functions, just because we can't trust the types - so its more effort to build stuff, or we don't add those checks and have increased risk of runtime errors.

This is obviously just one random example, the fact is that passing an empty string as a way to bypass a type validation is likely to introduce issues downstream. If we can't provide the value, in theory the function should allow not to provide that, or we're missing required data to call it.

It's like passing a Patient with an ID with empty string instead of an actual UUID as the ID. We use types to help us, when we bypass them, we add tech debt and increase the likelihood of bugs.

Reply:
Makes sense. Thank you! :)